### PR TITLE
Custom colors for protein structure

### PR DESCRIPTION
**Pull request by skhilton**
This PR adds the ability for the user to customize the colors on the protein structure.

The user can pick a color for the "data chains" (those specified in the data file) and the "other chains" (those not specified in the data file) using form entry fields.

@huddlej: Before we merge this we should test corner cases and update the docs. I can update the docs but if you could try to break things that would be great! We talked about two corner cases on the slack channel: the chains in the datafile do not exist in the pdb structure and what happens when no datafile is specified?

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -89,6 +96,9 @@ <h4>Instructions:</h4>
             <option value="surface">surface</option>
           </select>
           <button id="protein_plot_download" type="button" class="btn btn-light float-right"><img src="/assets/images/download.svg" /> Download PNG</button>
+          <label><input type="checkbox" id="colorCheckbox" name="colorCheckbox" value="colorChains"></input>color options</label>
+          <p><input id="protein-data-color" name='protein-data-color' text="text" class="form-control" placeholder="color for data chains" style="display: none"/></p>
+          <p><input id="protein-other-color" name='protein-other-color' text="text" class="form-control" placeholder="color for other chains" style="display: none"/></p>
```
Comment (huddlej) [inline index.html]: What if the “color options” toggle was styled more like a dropdown menu so users would be more likely to expect that they could click the color options to get a menu of settings? For example, we could use [the bootstrap dropdown class with a form inside to get the same effect](https://getbootstrap.com/docs/4.5/components/dropdowns/#forms).

File: index.html [hunk 2]
```diff
@@ -89,6 +96,9 @@ <h4>Instructions:</h4>
             <option value="surface">surface</option>
           </select>
           <button id="protein_plot_download" type="button" class="btn btn-light float-right"><img src="/assets/images/download.svg" /> Download PNG</button>
```
Comment (huddlej) [inline index.html:108]: I wish the "Download PNG" button was positioned to the right of the protein panel like the download buttons in the other panels.

File: assets/js/prot_struct.js [hunk 3]
```diff
@@ -6,6 +6,44 @@ stage.setParameters({
   backgroundColor: "white"
 });
 
+// color scheme
+function createProteinColorScheme(targetChains){
+  targetChains = ":" + targetChains.join(" or :");
```
Comment (huddlej) [inline assets/js/prot_struct.js:11]: If the user's data somehow produces a list of target chains with an empty string, this line produces an invalid query for NGL (I've included the full description in the main review text).

This line could be changed to something like:

```js
targetChains = ":" + targetChains.filter((item) => item != "").join(" or :");
```

**Review by huddlej — CHANGES_REQUESTED** [review]
This is generally working really well for me and I'm already loving the ability to explore where specific mutations are in flu relative to their trimeric position. There is one data-specific bug that I don't fully understand, but I've included instructions to recreate it below.

1. Load default view with flu data
2. Change PDB URL to HIV protein (https://raw.githubusercontent.com/dms-view/docs/master/_casestudy_data/5fyl_trimer_renumber.pdb)
3. Change data URL to HIV data (https://raw.githubusercontent.com/dms-view/docs/master/_casestudy_data/Dingens2019.csv)
4. Click “color options” checkbox and set the data color to “lightblue”. This produces the following error:

```
Uncaught TypeError: e.selectionList[i].test is not a function
    at i.e.atomColor (ngl.js:1)
    at i.Dd.atomColorToArray (ngl.js:1)
    at e (ngl.js:1)
    at u (ngl.js:1)
    at Ns.getColor (ngl.js:1)
    at ks.getSubdividedColor (ngl.js:1)
    at ngl.js:1
    at Ev.eachPolymer (ngl.js:1)
    at ngl.js:1
    at Iv.eachChain (ngl.js:1)
```

It looks like the problem is that the value of `targetChains` in the `createProteinColorScheme` function is `["G", "", "B"]`. If I filter out the empty string (with the code below), the coloring works kind of how I would expect, except that there are no “other chains” defined.

```js
targetChains = ":" + targetChains.filter((item) => item != "").join(" or :");
```

This seems like a bug with dms-view, but is it also maybe a bug with the input data? Should there ever be a case where users provide an empty string for a chain?

File: index.html (hunk 2 above)
Comment (skhilton) [reply index.html:108]: Yes! Consistency is good. Changed!

File: assets/js/prot_struct.js (hunk 3 above)
Comment (skhilton) [reply assets/js/prot_struct.js:11]: I don't believe there should ever be a empty string chain for the protein structure. I filtered out empty strings from the protein chains list when the list is created rather than each time a new color scheme is created.

**Review by huddlej — APPROVED** [review]

**Comment by huddlej** [comment]
I like this new interface a lot, @skhilton! I think it's ready to merge, if you do.

**Comment by skhilton** [comment]
@huddlej: I am going to merge this because I think it is working and it looks pretty good! I might play around trying to match the styles of the dropdown menus across all three panels but I don't think this is necessary for the merge.
